doc/wrap_native_functions.md

```diff
@@ -0,0 +1,156 @@
+
+```c++
+CAFFE2_API Tensor mm(const Tensor & self, const Tensor & mat2);
+```
+
+The method `Tensor::mm` is in `TenosrBody.h`, which, again is generated from
+native_functions.yaml. You can find it in
+`libtorch/include/ATen/core/TensorBody.h`.
+
+```c++
+class CAFFE2_API Tensor {
+  ...
+  Tensor mm(const Tensor & mat2) const;
+  ...
+};
+```
+
+Go programs can call C functions, via [Cgo](https://golang.org/cmd/cgo/).  For
+example, to call the standard C function exit with a parameter 0, we can write a
+Go code line.
+
+```go
+C.exit(0)
+```
+
+Here is a Go function MyExit that calls the C function exit.
+
+```go
+func MyExit(x int) {
+    C.exit(x)
+}
+```
+
+The Go function MM that calls the native function mm is as follows, in tensor.go.
+
+```go
+// MM multiplies each element of the input two tensors
+func MM(a, b Tensor) Tensor {
+	var t C.Tensor
+	MustNil(unsafe.Pointer(C.MM(C.Tensor(*a.T), C.Tensor(*b.T), &t)))
+	SetTensorFinalizer((*unsafe.Pointer)(&t))
+	return Tensor{(*unsafe.Pointer)(&t)}
+}
+```
+
+It takes two parameters a and b; both are of type Tensor.
+
+```go
+// Tensor wrappers a pointer to C.Tensor
+type Tensor struct {
+	T *unsafe.Pointer
+}
+```
+
+The Go type Tensor is indeed a pointer, which points to a C++ aten::Tenosr
+object. Go allows representing a C pointer by `unsafe.Pointer`.  We define `T`
+as `*unsafe.Pointer`, which is a Go pointer to a C pointer.  We need this
+pointer-to-pointer because only a Go pointer could have a finalizer attached to
+it.  And we need to attach a finalizer to each tensor object so that when the
+tensor is out-of-use, the finalizer can free the underlying C++ object.
+
+Also, we wrap this pointer-to-pointer in a struct, because Go doesn’t allow us
+to attach methods to a pointer type; however, the type Tensor needs methods like
+MM.
+
+With Cgo, all Go identifiers with prefix “C.” are C identifiers.  In this
+example, `C.MM` is the C function defined in `cgotorch/cgotorch.h`.
+
+```c
+const char *MM(Tensor a, Tensor b, Tensor *result);
+```
+
+You might be wondering why the return type is a string.  It is because all
+libtorch functions might throw C++ exceptions.  If it does, we want the C
+function MM to catch whatever thrown and convert the reason into a string, which
+is then converted by `MustNil`, called in the Go function `MM`, as shown in the
+above code snippet, into a Go panic.
+
+The C function MM has a C++ implementation in `cgotorch/torch.cc`.
+
+```c++
+const char *MM(Tensor a, Tensor b, Tensor *result) {
+  try {
+    at::Tensor c = at::mm(*a, *b);
+    *result = new at::Tensor(c);
+    return nullptr;
+  } catch (const std::exception &e) {
+    return exception_str(e.what());
+  }
+}
+```
+
+This C++ implementation is the one called by the Go version of MM under the name
+`C.MM`.  You might wonder why not Go MM calls the native function `mm` directly
+via `C.mm`, but `C.MM` in `cgotorch/torch.cc`.  The only reason is we need to
```

Comment:
The native functions have C++ declarations. My understanding is that we need provide C wrappers around C++ functions for go to call. So to answer 'why not Go MM calls the native function directly', besides exception handling, C++ objects need to be wrapped as C pointers for golang to manage their destruction explicitly. So we manually dereference C.Tensor to at::Tensor before calling c++ functions and manually copy the result into a heap tensor before returning.